pallets/liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,910 @@
+			if T::LiquidityMiningHandler::is_yield_farm_claimable(global_farm_id, yield_farm_id, amm_pool_id.clone()) {
+				//This should not fail on double claim, we need unclaimable_rewards
+				let fail_on_double_claim = false;
+				let (farm_id, reward_currency, claimed, unclaimable) = T::LiquidityMiningHandler::claim_rewards(
+					who.clone(),
+					deposit_id,
+					yield_farm_id,
+					fail_on_double_claim,
+				)?;
+
+				if !claimed.is_zero() {
+					Self::deposit_event(Event::RewardClaimed {
+						farm_id,
+						yield_farm_id,
+						who: who.clone(),
+						claimed,
+						reward_currency,
+					});
+				}
+
+				unclaimable_rewards = unclaimable;
+			}
+
+			let (farm_id, withdrawn_amount, is_destroyed) =
+				T::LiquidityMiningHandler::withdraw_lp_shares(deposit_id, yield_farm_id, unclaimable_rewards)?;
+
+			let lp_token = Self::get_lp_token(&amm_pool_id)?;
+			if !withdrawn_amount.is_zero() {
+				Self::deposit_event(Event::SharesWithdrawn {
+					farm_id,
+					yield_farm_id,
+					who: who.clone(),
+					lp_token,
+					amount: withdrawn_amount,
+				});
+			}
+
+			if is_destroyed {
+				Self::unlock_lp_tokens(amm_pool_id, who.clone(), withdrawn_amount)?;
+				T::NFTHandler::burn_from(&T::NftClassId::get(), &deposit_id)?;
+
+				Self::deposit_event(Event::DepositDestroyed {
+					who,
+					nft_instance_id: deposit_id,
+				});
+			}
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn account_id_for_all_lp_shares() -> AccountIdOf<T> {
+		<T as pallet::Config>::PalletId::get().into_account()
+	}
+
+	fn get_lp_token(amm_pool_id: &AccountIdOf<T>) -> Result<AssetId, Error<T>> {
+		let assets = T::AMM::get_pool_assets(amm_pool_id).ok_or(Error::<T>::CantGetAmmAssets)?;
+		let asset_pair = AssetPair::new(assets[0], assets[1]);
+
+		//NOTE: this check is important AMM:get_share_token() return `0` if amm doesn't exist
+		if !T::AMM::exists(asset_pair) {
+			return Err(Error::<T>::AmmPoolDoesNotExist);
+		}
+
+		Ok(T::AMM::get_share_token(asset_pair))
+	}
+
+	fn lock_lp_tokens(amm_pool_id: T::AccountId, who: T::AccountId, amount: Balance) -> Result<(), DispatchError> {
+		let lp_token = Self::get_lp_token(&amm_pool_id)?;
+
+		let service_account_for_lp_shares = Self::account_id_for_all_lp_shares();
+		MultiCurrencyOf::<T>::transfer(lp_token, &who, &service_account_for_lp_shares, amount)?;
+
+		Ok(())
+	}
+
+	fn unlock_lp_tokens(amm_pool_id: T::AccountId, who: T::AccountId, amount: Balance) -> Result<(), DispatchError> {
+		let lp_token = Self::get_lp_token(&amm_pool_id)?;
+
+		let service_account_for_lp_shares = Self::account_id_for_all_lp_shares();
+		MultiCurrencyOf::<T>::transfer(lp_token, &service_account_for_lp_shares, &who, amount)?;
+
+		Ok(())
+	}
+
+	fn mint_nft_representing_deposit(who: T::AccountId, deposit_id: &u128) -> Result<(), DispatchError> {
+		T::NFTHandler::mint_into(&T::NftClassId::get(), deposit_id, &who)?;
+
+		Ok(())
+	}
+
+	fn get_asset_balance_in_liquidity_mining_pool(
```

Comment:
```suggestion
	fn get_asset_balance_in_amm_pool(
```
This fn should return amount of incentivized asset in amm pool.